File: source/downloads/code/LeetCode/Java/3Sum.java
public class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
        List<List<Integer>> ret = new LinkedList<List<Integer>>();
        if (nums.length == 0)
            return ret;
        //　排序
        Arrays.sort(nums);
        
        for (int i=0;i<nums.length-2;i++){
            if(i==0 || nums[i] != nums[i-1]){
                int l = i+1;
                int r = nums.length-1;
                while (l<r){
                    int sums = nums[i]+nums[l]+nums[r];
                    if (sums==0){
                        // asList: array->list
                        ret.add(Arrays.asList(nums[i],nums[l],nums[r]));
                        while(l<r && nums[l]==nums[l+1]) l++;
                        while(l<r && nums[r]==nums[r-1]) r--;
                        l++;
                        r--;
                    }
                    else if(sums<0)
                        l ++;
                    else
                        r--;
                }
            }
        }
        return ret;
    }
}

File: downloads/code/LeetCode/Python/3Sum.py
class Solution(object):
    def threeSum(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """

        ret = []
        if not nums:
            return ret

        nums = sorted(nums)

        n = len(nums)
        for i in range(n):
            l = i+1
            r = n-1
            while l<r:
                val = nums[i]+nums[l]+nums[r]
                if val == 0:
                    ret.append(sorted([nums[l],nums[i],nums[r]]))
                    # should not break "-2,0,1,1,2"
                    l += 1
                elif val < 0:
                    l += 1
                else:
                    r -= 1
        return [list(t) for t in set(tuple(element) for element in ret)]
